{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5feae1f4-60a8-4f4f-b684-450bab191457",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "import warnings\n",
    "import re \n",
    "import argparse\n",
    "import subprocess\n",
    "import pandas as pd\n",
    "from pandas import isna \n",
    "import numpy as np\n",
    "\n",
    "from tempfile import NamedTemporaryFile\n",
    "\n",
    "from concurrent.futures import ThreadPoolExecutor, as_completed, ProcessPoolExecutor\n",
    "import concurrent.futures\n",
    "from scipy.cluster.hierarchy import fclusterdata\n",
    "from typing import Tuple\n",
    "from functools import partial\n",
    "import csv \n",
    "\n",
    "from typing import Dict\n",
    "\n",
    "### local imports \n",
    "sys.path.append(\"/home/150/as7425/rna_biogenesis_maps/read_classifier\")\n",
    "from bam_to_bed import bam_to_bed\n",
    "from process_genome import generate_genome_file, read_chromosomes_from_genome_file, filter_bed_by_chromosome_inplace\n",
    "from gtf_to_bed import gtf_to_bed, extend_gene_bed \n",
    "from parallel_bed_operations import split_bed_file, write_sorted_chunk\n",
    "from intersect import run_bedtools\n",
    "from parse_intersect import parse_output\n",
    "from process_read_end_positions import calculate_distance_for_unique_positions, calculate_distance_to_annotated_ends\n",
    "from process_gtf import parse_attributes, parse_gtf_line, get_gene_exon_table, get_biotypes \n",
    "from parse_classifications import parse_read_classification, summarize_error_cases, classify_read, print_classification_summary\n",
    "\n",
    "\n",
    "# set logging level \n",
    "import logging \n",
    "logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s', datefmt='%Y-%m-%d %H:%M:%S')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b1c25cfa-b101-40cd-b396-bdb9445b799a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# main \n",
    "def main(bam_file, gtf_file, output_table):\n",
    "\n",
    "    # fetch output directory \n",
    "    output_dir = os.path.dirname(output_table) \n",
    "\n",
    "    # generate a genome file for bedtools sort \n",
    "    genome_file = generate_genome_file(bam_file, output_dir) \n",
    "\n",
    "    # create a bed file of gene regions\n",
    "    gene_bed_file = gtf_to_bed(gtf_file, \"gene\", output_dir) \n",
    "\n",
    "    # filter for relevant chromsoomes\n",
    "    chromosomes = read_chromosomes_from_genome_file(genome_file)\n",
    "    filter_bed_by_chromosome_inplace(gene_bed_file, chromosomes)\n",
    "\n",
    "    # create a bed file of downstream of gene regions \n",
    "    dog_bed_file = extend_gene_bed(gene_bed_file, output_dir, genome_file) \n",
    "\n",
    "    # intersect reads against, exons, genes, and downstream of gene regions \n",
    "    intersect_files = run_bedtools(bam_file, gtf_file, dog_bed_file, genome_file, output_dir) \n",
    "\n",
    "    # parse the bedtools intersect files \n",
    "    df = parse_output(intersect_files[0], intersect_files[1], intersect_files[2], intersect_files[3], intersect_files[4])\n",
    "    \n",
    "    # Write df to a CSV file for testing\n",
    "    df.to_csv(output_dir + \"/test_df_end.csv\", index=False)\n",
    "    \n",
    "    # fetch distances to nearest transcript end sites \n",
    "    end_position_df = calculate_distance_to_annotated_ends(df, gtf_file, output_dir)\n",
    "\n",
    "    # save the output as CSV \n",
    "    end_position_df.to_csv(output_table, sep=\"\\t\", index=False)\n",
    "\n",
    "    # fetch the gene biotypes \n",
    "    biotypes = get_biotypes(gtf_file)\n",
    "\n",
    "    # classify each read \n",
    "    result_df = parse_read_classification(output_table, biotypes)\n",
    "\n",
    "    # print some summary statistics \n",
    "    print_classification_summary(result_df)\n",
    "\n",
    "    # Rename 'exon_count' to 'annotated_exon_count'\n",
    "    result_df.rename(columns={'exon_count': 'annotated_exon_count'}, inplace=True)\n",
    "\n",
    "    # Omit specific columns\n",
    "    columns_to_omit = ['read_end_strand', 'strand_sign', 'chromosome', 'position', 'strand', 'biotypes_info']\n",
    "    result_df.drop(columns=columns_to_omit, inplace=True)\n",
    "\n",
    "    # Save the table to output\n",
    "    result_df.to_csv(output_table, sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "537971fe-e197-4661-946f-74056c1d3e86",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-05-06 13:35:05 - INFO - Generating genome file\n",
      "2024-05-06 13:35:05 - INFO - Converting gene GTF to BED\n",
      "[E::idx_find_and_load] Could not retrieve index file for '/g/data/lf10/as7425/2023_mrna-biogenesis-maps/analysis/2024-03-28_ASR012_HeLa-total-RNA004-rep1/ASR012_HeLa-total-RNA004/classify_reads_tests/filtered_reads.bam'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    0         1         2                3   4  5\n",
      "0  12  56117967  56118467  ENSG00000229117 NaN  +\n",
      "{'1': 248956422, '10': 133797422, '11': 135086622, '12': 133275309, '13': 114364328, '14': 107043718, '15': 101991189, '16': 90338345, '17': 83257441, '18': 80373285, '19': 58617616, '2': 242193529, '20': 64444167, '21': 46709983, '22': 50818468, '3': 198295559, '4': 190214555, '5': 181538259, '6': 170805979, '7': 159345973, '8': 145138636, '9': 138394717, 'GL000008.2': 209709, 'GL000009.2': 201709, 'GL000194.1': 191469, 'GL000195.1': 182896, 'GL000205.2': 185591, 'GL000208.1': 92689, 'GL000213.1': 164239, 'GL000214.1': 137718, 'GL000216.2': 176608, 'GL000218.1': 161147, 'GL000219.1': 179198, 'GL000220.1': 161802, 'GL000221.1': 155397, 'GL000224.1': 179693, 'GL000225.1': 211173, 'GL000226.1': 15008, 'KI270302.1': 2274, 'KI270303.1': 1942, 'KI270304.1': 2165, 'KI270305.1': 1472, 'KI270310.1': 1201, 'KI270311.1': 12399, 'KI270312.1': 998, 'KI270315.1': 2276, 'KI270316.1': 1444, 'KI270317.1': 37690, 'KI270320.1': 4416, 'KI270322.1': 21476, 'KI270329.1': 1040, 'KI270330.1': 1652, 'KI270333.1': 2699, 'KI270334.1': 1368, 'KI270335.1': 1048, 'KI270336.1': 1026, 'KI270337.1': 1121, 'KI270338.1': 1428, 'KI270340.1': 1428, 'KI270362.1': 3530, 'KI270363.1': 1803, 'KI270364.1': 2855, 'KI270366.1': 8320, 'KI270371.1': 2805, 'KI270372.1': 1650, 'KI270373.1': 1451, 'KI270374.1': 2656, 'KI270375.1': 2378, 'KI270376.1': 1136, 'KI270378.1': 1048, 'KI270379.1': 1045, 'KI270381.1': 1930, 'KI270382.1': 4215, 'KI270383.1': 1750, 'KI270384.1': 1658, 'KI270385.1': 990, 'KI270386.1': 1788, 'KI270387.1': 1537, 'KI270388.1': 1216, 'KI270389.1': 1298, 'KI270390.1': 2387, 'KI270391.1': 1484, 'KI270392.1': 971, 'KI270393.1': 1308, 'KI270394.1': 970, 'KI270395.1': 1143, 'KI270396.1': 1880, 'KI270411.1': 2646, 'KI270412.1': 1179, 'KI270414.1': 2489, 'KI270417.1': 2043, 'KI270418.1': 2145, 'KI270419.1': 1029, 'KI270420.1': 2321, 'KI270422.1': 1445, 'KI270423.1': 981, 'KI270424.1': 2140, 'KI270425.1': 1884, 'KI270429.1': 1361, 'KI270435.1': 92983, 'KI270438.1': 112505, 'KI270442.1': 392061, 'KI270448.1': 7992, 'KI270465.1': 1774, 'KI270466.1': 1233, 'KI270467.1': 3920, 'KI270468.1': 4055, 'KI270507.1': 5353, 'KI270508.1': 1951, 'KI270509.1': 2318, 'KI270510.1': 2415, 'KI270511.1': 8127, 'KI270512.1': 22689, 'KI270515.1': 6361, 'KI270516.1': 1300, 'KI270517.1': 3253, 'KI270518.1': 2186, 'KI270519.1': 138126, 'KI270521.1': 7642, 'KI270522.1': 5674, 'KI270528.1': 2983, 'KI270529.1': 1899, 'KI270530.1': 2168, 'KI270538.1': 91309, 'KI270539.1': 993, 'KI270544.1': 1202, 'KI270548.1': 1599, 'KI270579.1': 31033, 'KI270580.1': 1553, 'KI270581.1': 7046, 'KI270582.1': 6504, 'KI270583.1': 1400, 'KI270584.1': 4513, 'KI270587.1': 2969, 'KI270588.1': 6158, 'KI270589.1': 44474, 'KI270590.1': 4685, 'KI270591.1': 5796, 'KI270593.1': 3041, 'KI270706.1': 175055, 'KI270707.1': 32032, 'KI270708.1': 127682, 'KI270709.1': 66860, 'KI270710.1': 40176, 'KI270711.1': 42210, 'KI270712.1': 176043, 'KI270713.1': 40745, 'KI270714.1': 41717, 'KI270715.1': 161471, 'KI270716.1': 153799, 'KI270717.1': 40062, 'KI270718.1': 38054, 'KI270719.1': 176845, 'KI270720.1': 39050, 'KI270721.1': 100316, 'KI270722.1': 194050, 'KI270723.1': 38115, 'KI270724.1': 39555, 'KI270725.1': 172810, 'KI270726.1': 43739, 'KI270727.1': 448248, 'KI270728.1': 1872759, 'KI270729.1': 280839, 'KI270730.1': 112551, 'KI270731.1': 150754, 'KI270732.1': 41543, 'KI270733.1': 179772, 'KI270734.1': 165050, 'KI270735.1': 42811, 'KI270736.1': 181920, 'KI270737.1': 103838, 'KI270738.1': 99375, 'KI270739.1': 73985, 'KI270740.1': 37240, 'KI270741.1': 157432, 'KI270742.1': 186739, 'KI270743.1': 210658, 'KI270744.1': 168472, 'KI270745.1': 41891, 'KI270746.1': 66486, 'KI270747.1': 198735, 'KI270748.1': 93321, 'KI270749.1': 158759, 'KI270750.1': 148850, 'KI270751.1': 150742, 'KI270752.1': 27745, 'KI270753.1': 62944, 'KI270754.1': 40191, 'KI270755.1': 36723, 'KI270756.1': 79590, 'KI270757.1': 71251, 'MT': 16569, 'X': 156040895, 'Y': 57227415}\n",
      "[8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 1/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 2/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 3/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 4/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 5/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 6/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 7/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 8/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 9/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 10/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 11/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 12/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 13/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 14/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 15/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 16/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 17/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 43/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 19/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 20/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 21/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 22/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 23/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 24/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 25/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 26/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 27/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 28/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 29/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 30/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 31/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 32/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 33/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 34/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 35/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 36/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 37/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 38/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 39/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 40/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 41/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 42/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 44/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 18/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 45/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 46/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 47/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 48/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 49/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 50/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 51/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 52/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 53/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 54/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 55/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 56/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 57/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 58/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 59/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 60/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 61/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 62/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 63/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 64/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 65/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 66/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 67/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 68/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 69/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 70/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 71/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 72/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 73/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 74/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 75/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 76/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 77/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 78/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 79/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 80/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 81/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 82/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 83/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 84/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 85/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 86/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 87/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 88/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 89/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 90/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 91/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 92/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 93/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 94/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 95/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 96/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 97/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 98/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 99/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 100/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 101/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 102/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 103/104\n",
      "2024-05-06 13:35:06 - INFO - Converting BAM to BED: 104/104\n",
      "2024-05-06 13:35:07 - INFO - Running: Merging and sorting all BED files\n",
      "2024-05-06 13:35:07 - INFO - Calculating alignment lengths\n",
      "2024-05-06 13:35:07 - INFO - Calculating sequence lengths\n",
      "2024-05-06 13:35:08 - INFO - Converting exon GTF to BED\n",
      "2024-05-06 13:35:08 - INFO - Converting gene GTF to BED\n",
      "2024-05-06 13:35:08 - INFO - Calculating alignment overlap with exons, genes and DOG regions\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    0         1         2                3   4  5\n",
      "0  12  56117967  56118467  ENSG00000229117 NaN  +\n",
      "{'1': 248956422, '10': 133797422, '11': 135086622, '12': 133275309, '13': 114364328, '14': 107043718, '15': 101991189, '16': 90338345, '17': 83257441, '18': 80373285, '19': 58617616, '2': 242193529, '20': 64444167, '21': 46709983, '22': 50818468, '3': 198295559, '4': 190214555, '5': 181538259, '6': 170805979, '7': 159345973, '8': 145138636, '9': 138394717, 'GL000008.2': 209709, 'GL000009.2': 201709, 'GL000194.1': 191469, 'GL000195.1': 182896, 'GL000205.2': 185591, 'GL000208.1': 92689, 'GL000213.1': 164239, 'GL000214.1': 137718, 'GL000216.2': 176608, 'GL000218.1': 161147, 'GL000219.1': 179198, 'GL000220.1': 161802, 'GL000221.1': 155397, 'GL000224.1': 179693, 'GL000225.1': 211173, 'GL000226.1': 15008, 'KI270302.1': 2274, 'KI270303.1': 1942, 'KI270304.1': 2165, 'KI270305.1': 1472, 'KI270310.1': 1201, 'KI270311.1': 12399, 'KI270312.1': 998, 'KI270315.1': 2276, 'KI270316.1': 1444, 'KI270317.1': 37690, 'KI270320.1': 4416, 'KI270322.1': 21476, 'KI270329.1': 1040, 'KI270330.1': 1652, 'KI270333.1': 2699, 'KI270334.1': 1368, 'KI270335.1': 1048, 'KI270336.1': 1026, 'KI270337.1': 1121, 'KI270338.1': 1428, 'KI270340.1': 1428, 'KI270362.1': 3530, 'KI270363.1': 1803, 'KI270364.1': 2855, 'KI270366.1': 8320, 'KI270371.1': 2805, 'KI270372.1': 1650, 'KI270373.1': 1451, 'KI270374.1': 2656, 'KI270375.1': 2378, 'KI270376.1': 1136, 'KI270378.1': 1048, 'KI270379.1': 1045, 'KI270381.1': 1930, 'KI270382.1': 4215, 'KI270383.1': 1750, 'KI270384.1': 1658, 'KI270385.1': 990, 'KI270386.1': 1788, 'KI270387.1': 1537, 'KI270388.1': 1216, 'KI270389.1': 1298, 'KI270390.1': 2387, 'KI270391.1': 1484, 'KI270392.1': 971, 'KI270393.1': 1308, 'KI270394.1': 970, 'KI270395.1': 1143, 'KI270396.1': 1880, 'KI270411.1': 2646, 'KI270412.1': 1179, 'KI270414.1': 2489, 'KI270417.1': 2043, 'KI270418.1': 2145, 'KI270419.1': 1029, 'KI270420.1': 2321, 'KI270422.1': 1445, 'KI270423.1': 981, 'KI270424.1': 2140, 'KI270425.1': 1884, 'KI270429.1': 1361, 'KI270435.1': 92983, 'KI270438.1': 112505, 'KI270442.1': 392061, 'KI270448.1': 7992, 'KI270465.1': 1774, 'KI270466.1': 1233, 'KI270467.1': 3920, 'KI270468.1': 4055, 'KI270507.1': 5353, 'KI270508.1': 1951, 'KI270509.1': 2318, 'KI270510.1': 2415, 'KI270511.1': 8127, 'KI270512.1': 22689, 'KI270515.1': 6361, 'KI270516.1': 1300, 'KI270517.1': 3253, 'KI270518.1': 2186, 'KI270519.1': 138126, 'KI270521.1': 7642, 'KI270522.1': 5674, 'KI270528.1': 2983, 'KI270529.1': 1899, 'KI270530.1': 2168, 'KI270538.1': 91309, 'KI270539.1': 993, 'KI270544.1': 1202, 'KI270548.1': 1599, 'KI270579.1': 31033, 'KI270580.1': 1553, 'KI270581.1': 7046, 'KI270582.1': 6504, 'KI270583.1': 1400, 'KI270584.1': 4513, 'KI270587.1': 2969, 'KI270588.1': 6158, 'KI270589.1': 44474, 'KI270590.1': 4685, 'KI270591.1': 5796, 'KI270593.1': 3041, 'KI270706.1': 175055, 'KI270707.1': 32032, 'KI270708.1': 127682, 'KI270709.1': 66860, 'KI270710.1': 40176, 'KI270711.1': 42210, 'KI270712.1': 176043, 'KI270713.1': 40745, 'KI270714.1': 41717, 'KI270715.1': 161471, 'KI270716.1': 153799, 'KI270717.1': 40062, 'KI270718.1': 38054, 'KI270719.1': 176845, 'KI270720.1': 39050, 'KI270721.1': 100316, 'KI270722.1': 194050, 'KI270723.1': 38115, 'KI270724.1': 39555, 'KI270725.1': 172810, 'KI270726.1': 43739, 'KI270727.1': 448248, 'KI270728.1': 1872759, 'KI270729.1': 280839, 'KI270730.1': 112551, 'KI270731.1': 150754, 'KI270732.1': 41543, 'KI270733.1': 179772, 'KI270734.1': 165050, 'KI270735.1': 42811, 'KI270736.1': 181920, 'KI270737.1': 103838, 'KI270738.1': 99375, 'KI270739.1': 73985, 'KI270740.1': 37240, 'KI270741.1': 157432, 'KI270742.1': 186739, 'KI270743.1': 210658, 'KI270744.1': 168472, 'KI270745.1': 41891, 'KI270746.1': 66486, 'KI270747.1': 198735, 'KI270748.1': 93321, 'KI270749.1': 158759, 'KI270750.1': 148850, 'KI270751.1': 150742, 'KI270752.1': 27745, 'KI270753.1': 62944, 'KI270754.1': 40191, 'KI270755.1': 36723, 'KI270756.1': 79590, 'KI270757.1': 71251, 'MT': 16569, 'X': 156040895, 'Y': 57227415}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/150/as7425/rna_biogenesis_maps/read_classifier/parse_intersect.py:15: UserWarning: Warning: Essential file /g/data/lf10/as7425/2023_mrna-biogenesis-maps/analysis/2024-03-28_ASR012_HeLa-total-RNA004-rep1/ASR012_HeLa-total-RNA004/classify_reads_tests/out/tmpxfawa2ps_final.bed_dog_overlap_final.bed is empty. Some data may not be processed.\n",
      "  warnings.warn(f\"Warning: Essential file {file} is empty. Some data may not be processed.\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Preview of exon_overlap_group:\n",
      "                                read_id     exon_gene_id  exon_base_overlap\n",
      "0  001e8a5e-5cce-4150-957a-1a0d269f5f31  ENSG00000229117                423\n",
      "1  0022dddc-afa7-4458-96f9-3607ab6f62f3  ENSG00000229117                231\n",
      "2  007ec063-7ba1-466d-9c17-64cf25ffa0ba  ENSG00000229117                455\n",
      "3  0084ef5d-6e3d-4583-aa05-4bdf6345a592  ENSG00000229117                454\n",
      "4  0090efa6-f042-41b0-80ff-3f5e6cfe4730  ENSG00000229117                380\n",
      "Preview of gene_df_filtered:\n",
      "   read_chrom  read_fragment_start  read_fragment_end  \\\n",
      "0          12             56116554           56116631   \n",
      "1          12             56116560           56116623   \n",
      "2          12             56116574           56116659   \n",
      "3          12             56116588           56116629   \n",
      "4          12             56116596           56116603   \n",
      "\n",
      "                                read_id read-alignment_length read_strand  \\\n",
      "0  74e5758a-0d72-455b-913d-2c2a6849c761             630,561,3           +   \n",
      "1  120fbca5-7836-4e12-8649-c05570c6e9b0             561,474,3           +   \n",
      "2  bf722428-7d60-4998-b0bf-a8c0e5882931             562,474,3           +   \n",
      "3  fa0abf77-a36b-42d5-aad7-538529b07ac9             556,464,3           +   \n",
      "4  fb1bf436-f133-409b-9102-227b8a4bbff9             535,443,3           +   \n",
      "\n",
      "   gene_chrom  gene_start  gene_end          gene_id  gene_biotype  \\\n",
      "0          12    56116589  56117967  ENSG00000229117           NaN   \n",
      "1          12    56116589  56117967  ENSG00000229117           NaN   \n",
      "2          12    56116589  56117967  ENSG00000229117           NaN   \n",
      "3          12    56116589  56117967  ENSG00000229117           NaN   \n",
      "4          12    56116589  56117967  ENSG00000229117           NaN   \n",
      "\n",
      "  gene_strand  gene_base_overlap  \n",
      "0           +                 42  \n",
      "1           +                 34  \n",
      "2           +                 70  \n",
      "3           +                 40  \n",
      "4           +                  7  \n",
      "Preview of gene_overlap_sum:\n",
      "                                read_id          gene_id  \\\n",
      "0  001e8a5e-5cce-4150-957a-1a0d269f5f31  ENSG00000229117   \n",
      "1  0022dddc-afa7-4458-96f9-3607ab6f62f3  ENSG00000229117   \n",
      "2  007ec063-7ba1-466d-9c17-64cf25ffa0ba  ENSG00000229117   \n",
      "3  0084ef5d-6e3d-4583-aa05-4bdf6345a592  ENSG00000229117   \n",
      "4  0090efa6-f042-41b0-80ff-3f5e6cfe4730  ENSG00000229117   \n",
      "\n",
      "  read-alignment_length  gene_base_overlap     exon_gene_id  exon_base_overlap  \n",
      "0             514,423,2                423  ENSG00000229117                423  \n",
      "1             313,231,0                231  ENSG00000229117                231  \n",
      "2             537,455,2                455  ENSG00000229117                455  \n",
      "3             540,454,2                454  ENSG00000229117                454  \n",
      "4             464,380,2                380  ENSG00000229117                380  \n",
      "Calculating DOG coverage\n",
      "DOG data is empty, defaulting to empty DataFrame.\n",
      "Preview of dog_overlap_group:\n",
      "Empty DataFrame\n",
      "Columns: [read_id, dog_gene_id, dog_base_overlap]\n",
      "Index: []\n",
      "Preview of dog_df:\n",
      "Empty DataFrame\n",
      "Columns: [read_chrom, read_fragment_start, read_fragment_end, read_id, read-alignment_length, read_strand, gene_chrom, gene_start, gene_end, dog_gene_id, gene_biotype, gene_strand, dog_base_overlap]\n",
      "Index: []\n",
      "Preview of dog_overlap_group:\n",
      "Empty DataFrame\n",
      "Columns: [dog_base_overlap]\n",
      "Index: []\n",
      "Preview of dog_overlap_group:\n",
      "Empty DataFrame\n",
      "Columns: [dog_base_overlap]\n",
      "Index: []\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-05-06 13:35:11 - INFO - Converting transcript GTF to BED\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Preview of dog_overlap_filtered:\n",
      "Empty DataFrame\n",
      "Columns: [index, dog_gene_id, dog_base_overlap, read_id, gene_id]\n",
      "Index: []\n",
      "Preview of gene_overlap_sum after merge\n",
      "                                read_id read-alignment_length  \\\n",
      "0  0b3e2e76-d3de-4d9a-a86d-d1d09dc4df60             483,409,3   \n",
      "1  4778ced8-b519-4f4c-837a-d9b7ba62abd4             611,492,1   \n",
      "2  3874fc97-c889-460e-b180-c3ea7662dacc             515,443,2   \n",
      "3  df69a10b-af0d-428c-ac52-883ad4d536f5             542,444,2   \n",
      "4  1e9b6318-5acc-44cc-afb5-7d87263a51f7             564,460,2   \n",
      "\n",
      "           gene_id  gene_base_overlap     exon_gene_id  exon_base_overlap  \\\n",
      "0  ENSG00000229117                403  ENSG00000229117                403   \n",
      "1  ENSG00000229117                492  ENSG00000229117                492   \n",
      "2  ENSG00000229117                443  ENSG00000229117                443   \n",
      "3  ENSG00000229117                444  ENSG00000229117                444   \n",
      "4  ENSG00000229117                460  ENSG00000229117                460   \n",
      "\n",
      "  DOG_overlap  \n",
      "0         NaN  \n",
      "1         NaN  \n",
      "2         NaN  \n",
      "3         NaN  \n",
      "4         NaN  \n",
      "Number of 'NA' values in end coordinates:  0\n",
      "An error occurred: 'read_end_chromosome'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-05-06 13:35:11 - INFO - Starting to parse GTF file.\n",
      "2024-05-06 13:35:11 - INFO - GTF file parsed. Processing data.\n",
      "2024-05-06 13:35:11 - INFO - Data processed successfully. Returning final DataFrame.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "An error occurred while calculating distances: No objects to concatenate\n",
      "head of the gtf DataFrame:\n",
      "    seqname          source feature     start       end score strand frame  \\\n",
      "0       12  ensembl_havana    gene  56116590  56117967     .      +     .   \n",
      "\n",
      "                                           attribute  \\\n",
      "0  gene_id \"ENSG00000229117\"; gene_version \"9\"; g...   \n",
      "\n",
      "                                          attributes          gene_id  \\\n",
      "0  {'gene_id': 'ENSG00000229117', 'gene_version':...  ENSG00000229117   \n",
      "\n",
      "     gene_biotype  \n",
      "0  protein_coding  \n",
      "\n",
      "summary stats:\n",
      "number of unique genes:  1\n",
      "number of unique gene biotypes:  1\n"
     ]
    },
    {
     "ename": "EmptyDataError",
     "evalue": "No columns to parse from file",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mEmptyDataError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[3], line 8\u001b[0m\n\u001b[1;32m      4\u001b[0m gtf_file \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m/g/data/lf10/as7425/2023_mrna-biogenesis-maps/analysis/2024-03-28_ASR012_HeLa-total-RNA004-rep1/ASR012_HeLa-total-RNA004/classify_reads_tests/gene_annotation.gtf\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m      6\u001b[0m output_table \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m/g/data/lf10/as7425/2023_mrna-biogenesis-maps/analysis/2024-03-28_ASR012_HeLa-total-RNA004-rep1/ASR012_HeLa-total-RNA004/classify_reads_tests/out/read_classification.txt\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m----> 8\u001b[0m \u001b[43mmain\u001b[49m\u001b[43m(\u001b[49m\u001b[43mbam_file\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgtf_file\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43moutput_table\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[2], line 90\u001b[0m, in \u001b[0;36mmain\u001b[0;34m(bam_file, gtf_file, output_table)\u001b[0m\n\u001b[1;32m     87\u001b[0m biotypes \u001b[38;5;241m=\u001b[39m get_biotypes(gtf_file)\n\u001b[1;32m     89\u001b[0m \u001b[38;5;66;03m# classify each read \u001b[39;00m\n\u001b[0;32m---> 90\u001b[0m result_df \u001b[38;5;241m=\u001b[39m \u001b[43mparse_read_classification\u001b[49m\u001b[43m(\u001b[49m\u001b[43moutput_table\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbiotypes\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     92\u001b[0m \u001b[38;5;66;03m# print some summary statistics \u001b[39;00m\n\u001b[1;32m     93\u001b[0m print_classification_summary(result_df)\n",
      "File \u001b[0;32m~/rna_biogenesis_maps/read_classifier/parse_classifications.py:25\u001b[0m, in \u001b[0;36mparse_read_classification\u001b[0;34m(file_path, biotypes)\u001b[0m\n\u001b[1;32m     22\u001b[0m sample_name \u001b[38;5;241m=\u001b[39m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mbasename(file_path)\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m_\u001b[39m\u001b[38;5;124m'\u001b[39m)[\u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m     24\u001b[0m \u001b[38;5;66;03m# read the data\u001b[39;00m\n\u001b[0;32m---> 25\u001b[0m df \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfile_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlow_memory\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[1;32m     27\u001b[0m \u001b[38;5;66;03m# Add biotype, gene_name, and exon_count information\u001b[39;00m\n\u001b[1;32m     28\u001b[0m df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mbiotypes_info\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgene_id\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mmap(biotypes)\n",
      "File \u001b[0;32m/g/data/lf10/as7425/apps/miniconda3/lib/python3.9/site-packages/pandas/io/parsers.py:610\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, dialect, error_bad_lines, warn_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options)\u001b[0m\n\u001b[1;32m    605\u001b[0m kwds_defaults \u001b[38;5;241m=\u001b[39m _refine_defaults_read(\n\u001b[1;32m    606\u001b[0m     dialect, delimiter, delim_whitespace, engine, sep, defaults\u001b[38;5;241m=\u001b[39m{\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdelimiter\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m,\u001b[39m\u001b[38;5;124m\"\u001b[39m}\n\u001b[1;32m    607\u001b[0m )\n\u001b[1;32m    608\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m--> 610\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/g/data/lf10/as7425/apps/miniconda3/lib/python3.9/site-packages/pandas/io/parsers.py:462\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    459\u001b[0m _validate_names(kwds\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[1;32m    461\u001b[0m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[0;32m--> 462\u001b[0m parser \u001b[38;5;241m=\u001b[39m \u001b[43mTextFileReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    464\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n\u001b[1;32m    465\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n",
      "File \u001b[0;32m/g/data/lf10/as7425/apps/miniconda3/lib/python3.9/site-packages/pandas/io/parsers.py:819\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m    816\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01min\u001b[39;00m kwds:\n\u001b[1;32m    817\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[0;32m--> 819\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_engine\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/g/data/lf10/as7425/apps/miniconda3/lib/python3.9/site-packages/pandas/io/parsers.py:1050\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[0;34m(self, engine)\u001b[0m\n\u001b[1;32m   1046\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m   1047\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mUnknown engine: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mengine\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m (valid options are \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mmapping\u001b[38;5;241m.\u001b[39mkeys()\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m)\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1048\u001b[0m     )\n\u001b[1;32m   1049\u001b[0m \u001b[38;5;66;03m# error: Too many arguments for \"ParserBase\"\u001b[39;00m\n\u001b[0;32m-> 1050\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mmapping\u001b[49m\u001b[43m[\u001b[49m\u001b[43mengine\u001b[49m\u001b[43m]\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/g/data/lf10/as7425/apps/miniconda3/lib/python3.9/site-packages/pandas/io/parsers.py:1898\u001b[0m, in \u001b[0;36mCParserWrapper.__init__\u001b[0;34m(self, src, **kwds)\u001b[0m\n\u001b[1;32m   1895\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles\u001b[38;5;241m.\u001b[39mhandle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles\u001b[38;5;241m.\u001b[39mhandle\u001b[38;5;241m.\u001b[39mmmap  \u001b[38;5;66;03m# type: ignore[union-attr]\u001b[39;00m\n\u001b[1;32m   1897\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1898\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_reader \u001b[38;5;241m=\u001b[39m \u001b[43mparsers\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mTextReader\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhandles\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhandle\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1899\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n\u001b[1;32m   1900\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles\u001b[38;5;241m.\u001b[39mclose()\n",
      "File \u001b[0;32mpandas/_libs/parsers.pyx:521\u001b[0m, in \u001b[0;36mpandas._libs.parsers.TextReader.__cinit__\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mEmptyDataError\u001b[0m: No columns to parse from file"
     ]
    }
   ],
   "source": [
    "# run on a subset of data \n",
    "\n",
    "# filtered read subsets from ASR012 \n",
    "bam_file = \"/g/data/lf10/as7425/2023_mrna-biogenesis-maps/analysis/2024-03-28_ASR012_HeLa-total-RNA004-rep1/ASR012_HeLa-total-RNA004/classify_reads_tests/filtered_reads.bam\"\n",
    "\n",
    "gtf_file = \"/g/data/lf10/as7425/2023_mrna-biogenesis-maps/analysis/2024-03-28_ASR012_HeLa-total-RNA004-rep1/ASR012_HeLa-total-RNA004/classify_reads_tests/gene_annotation.gtf\"\n",
    "\n",
    "output_table = \"/g/data/lf10/as7425/2023_mrna-biogenesis-maps/analysis/2024-03-28_ASR012_HeLa-total-RNA004-rep1/ASR012_HeLa-total-RNA004/classify_reads_tests/out/read_classification.txt\"\n",
    "\n",
    "main(bam_file, gtf_file, output_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f40746ec-4e88-4283-be30-8b908e679424",
   "metadata": {},
   "outputs": [],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    \n",
    "    parser = argparse.ArgumentParser(description='Classify chromatin-bound direct RNA-Seq reads for splicing and transcriptional status')\n",
    "    parser.add_argument('-b', '--bam_file', type=str, help='Path to the input bam file.')\n",
    "    parser.add_argument('-g', '--gtf_file', type=str, help='Path to the GTF file.')\n",
    "    parser.add_argument('-o', '--output_table', type=str, default=None, help='Optional path to the output file.')\n",
    "\n",
    "    args = parser.parse_args()\n",
    "\n",
    "    bam_file = args.bam_file\n",
    "    gtf_file = args.gtf_file \n",
    "    output_table = args.output_table\n",
    "\n",
    "    main(bam_file, gtf_file, output_table)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1acea99d-5c01-46a9-b12c-1dc003fc38a8",
   "metadata": {},
   "source": [
    "###### bam_file = \"/g/data/lf10/as7425/2023_mrna-biogenesis-maps/analysis/2024-03-20_ASR011_HeLa-POINT-RNA004-rep1/ASR011_HeLa-POINT-RNA004-rep1_primary_genome_alignments_modCalled.bam\"\n",
    "gtf_file = \"/g/data/lf10/as7425/genomes/human_genome/Homo_sapiens_transcriptsOnly_GRCh38.104.chr.gtf\"\n",
    "output_table = \"/g/data/lf10/as7425/2023_mrna-biogenesis-maps/analysis/2024-03-20_ASR011_HeLa-POINT-RNA004-rep1/classify_reads/read_classification.txt\"\n",
    "\n",
    "main(bam_file, gtf_file, output_table)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
